Change Behaviour of Provider Configuration to Only Use auth_login when No Token is Provided

marcboudreau opened the pull request:
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes #1157

I believe this PR would offer a viable solution to the problem raised in Issue #1157, by allowing a Terraform configuration to contain an **auth_login** block in the **vault** provider configuration, but the provider would use a Vault Token found in the environment instead of using the **auth_login** block.

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
provider: Change behaviour to only use auth_login block when no Vault Token is present in the environment
```

Output from acceptance testing:

```
$ TESTARGS="--run AuthLoginProviderConfigure" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test -v --run AuthLoginProviderConfigure -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/codegen   0.390s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode--- 1.610s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode--- 1.144s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet--- 0.820s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role------- 1.907s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template--- 1.380s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation      0.511s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
?       github.com/hashicorp/terraform-provider-vault/testutil  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/util      0.602s [no tests to run]
=== RUN   TestAccAuthLoginProviderConfigure
--- PASS: TestAccAuthLoginProviderConfigure (24.29s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     26.415s
```

RobertYoung commented:
Is this something that Hashicorp will consider? We have a use case where we want to use jwt in CI but tokens for development

marcboudreau commented:
Apologies for the unnecessary closing and reopening of this issue.  I've rebased my branch and adjusted my approach based on the refactoring of _provider.go_.
